File: scripts/fetch-solar-data.mjs
// Fetches solar irradiance climatology from NASA POWER regional API
// Gets entire latitude bands in single requests — much faster than point-by-point
//
// Run: node scripts/fetch-solar-data.mjs
// Output: public/solar-data.json

import { writeFileSync } from 'fs'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __dirname = dirname(fileURLToPath(import.meta.url))
const OUT_PATH = join(__dirname, '..', 'public', 'solar-data.json')

// NASA POWER regional API supports max 10° lat × full lon bands
const LON_MIN = -180, LON_MAX = 180
const LAT_BAND = 10
const DELAY = 2000

console.log(`\n  NASA POWER Solar Data Downloader (Regional API)`)
console.log(`  ─────────────────────────────────────────────────`)

const bands = []
for (let lat = -90; lat < 90; lat += LAT_BAND) {
  bands.push({ latMin: lat, latMax: Math.min(lat + LAT_BAND, 90) })
}

console.log(`  Fetching ${bands.length} latitude bands...`)
console.log(`  Source: CERES SYN1deg / MERRA-2 (2001-2020 climatology)\n`)

const allPoints = []
const t0 = Date.now()

for (let i = 0; i < bands.length; i++) {
  const { latMin, latMax } = bands[i]
  const pct = ((i + 1) / bands.length * 100).toFixed(0)
  process.stdout.write(`  [${pct}%] Band ${i + 1}/${bands.length}: lat ${latMin}° to ${latMax}°...`)

  const url = `https://power.larc.nasa.gov/api/temporal/climatology/regional?parameters=ALLSKY_SFC_SW_DWN&community=RE&latitude-min=${latMin}&latitude-max=${latMax}&longitude-min=${LON_MIN}&longitude-max=${LON_MAX}&format=JSON&start=2001&end=2020`

  try {
    const resp = await fetch(url)
    if (!resp.ok) {
      console.log(` HTTP ${resp.status}`)
      if (resp.status === 429) {
        console.log(`  ⏳ Rate limited, waiting 30s...`)
        await sleep(30000)
        i--
        continue
      }
      continue
    }

    const data = await resp.json()

    // Regional API returns features with geometry and properties
    const features = data?.features
    if (features && features.length > 0) {
      for (const feat of features) {
        const coords = feat.geometry?.coordinates
        const ann = feat.properties?.parameter?.ALLSKY_SFC_SW_DWN?.ANN
        if (coords && ann !== undefined && ann > 0) {
          // coords = [lon, lat]
          const hours = Math.round(ann * 365 * 0.75)
          allPoints.push([coords[1], coords[0], hours])
        }
      }
      console.log(` ✓ ${features.length} points`)
    } else {
      // Maybe different response format — try parsing as grid
      const params = data?.properties?.parameter?.ALLSKY_SFC_SW_DWN
      if (params) {
        // Single region value
        const ann = params.ANN
        if (ann > 0) {
          const midLat = (latMin + latMax) / 2
          for (let lon = LON_MIN; lon < LON_MAX; lon += 3) {
            allPoints.push([midLat, lon, Math.round(ann * 365 * 0.75)])
          }
        }
        console.log(` ✓ (aggregated)`)
      } else {
        console.log(` ✗ no data`)
      }
    }
  } catch (e) {
    console.log(` ✗ ${e.message}`)
  }

  if (i < bands.length - 1) await sleep(DELAY)
}

console.log(`\n  Received ${allPoints.length} points total`)

// If regional API didn't give enough granularity, fall back to point queries for a coarser grid
if (allPoints.length < 500) {
  console.log(`  Regional API returned insufficient data, falling back to point queries...`)
  console.log(`  Using 5° grid for speed...\n`)

  const STEP = 5
  const lats = [], lons = []
  for (let lat = -80; lat <= 80; lat += STEP) lats.push(lat)
  for (let lon = -180; lon < 180; lon += STEP) lons.push(lon)

  const total = lats.length * lons.length
  let done = 0

  for (const lat of lats) {
    for (const lon of lons) {
      done++
      const url = `https://power.larc.nasa.gov/api/temporal/climatology/point?parameters=ALLSKY_SFC_SW_DWN&community=RE&longitude=${lon}&latitude=${lat}&format=JSON&start=2001&end=2020`

      try {
        const resp = await fetch(url)
        if (resp.ok) {
          const data = await resp.json()
          const ann = data?.properties?.parameter?.ALLSKY_SFC_SW_DWN?.ANN
          if (ann > 0) {
            allPoints.push([lat, lon, Math.round(ann * 365 * 0.75)])
          }
        } else if (resp.status === 429) {
          await sleep(15000)
          done--
          continue
        }
      } catch {}

      if (done % 20 === 0) {
        const pct = (done / total * 100).toFixed(1)
        const elapsed = (Date.now() - t0) / 1000
        const eta = Math.ceil((total - done) * (elapsed / done))
        process.stdout.write(`\r  ${pct}% (${done}/${total}) ETA: ${Math.floor(eta/60)}m${eta%60}s   `)
      }

      await sleep(400)
    }
  }
  console.log('')
}

// Compute stats & save
const hours = allPoints.map(p => p[2])
const min = Math.min(...hours)
const max = Math.max(...hours)
const avg = Math.round(hours.reduce((a, b) => a + b, 0) / hours.length)

// Determine step from data
const uniqueLats = [...new Set(allPoints.map(p => p[0]))].sort((a, b) => a - b)
const step = uniqueLats.length > 1 ? Math.round(uniqueLats[1] - uniqueLats[0]) : 5

const output = {
  step,
  year: '2001-2020 climatology',
  source: 'NASA POWER (CERES SYN1deg / MERRA-2)',
  stats: { min, max, avg, count: allPoints.length, total: allPoints.length },
  grid: allPoints,
}

writeFileSync(OUT_PATH, JSON.stringify(output))

const elapsed = ((Date.now() - t0) / 60000).toFixed(1)
console.log(`\n  ✓ Done! ${allPoints.length} points`)
console.log(`  ✓ Range: ${min}–${max} sunshine-equivalent hours/year (avg: ${avg})`)
console.log(`  ✓ Saved to public/solar-data.json`)
console.log(`  ✓ Elapsed: ${elapsed} minutes\n`)

function sleep(ms) { return new Promise(r => setTimeout(r, ms)) }

File: scripts/fetch-solar-data.py
"""
Fetch global solar irradiance from NASA POWER API using parallel requests.
Data: CERES SYN1deg / MERRA-2, 20-year climatology (2001-2020)

Run: python scripts/fetch-solar-data.py
Output: public/solar-data.json
"""

import json
import sys
import time
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

STEP = 5  # degrees
WORKERS = 8  # parallel requests
OUT = Path(__file__).parent.parent / "public" / "solar-data.json"

# Build grid
lats = list(range(-80, 81, STEP))
lons = list(range(-180, 180, STEP))
coords = [(lat, lon) for lat in lats for lon in lons]
total = len(coords)

print(f"\n  NASA POWER Solar Data (parallel)")
print(f"  --------------------------------")
print(f"  Grid: {STEP}deg => {len(lats)}x{len(lons)} = {total} points")
print(f"  Workers: {WORKERS}")
print(f"  Source: CERES/MERRA-2 climatology (2001-2020)\n")


def fetch_point(lat, lon):
    url = (
        f"https://power.larc.nasa.gov/api/temporal/climatology/point?"
        f"parameters=ALLSKY_SFC_SW_DWN&community=RE"
        f"&longitude={lon}&latitude={lat}&format=JSON&start=2001&end=2020"
    )
    for attempt in range(3):
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "solar-map/1.0"})
            with urllib.request.urlopen(req, timeout=30) as resp:
                data = json.loads(resp.read())
                ann = data.get("properties", {}).get("parameter", {}).get("ALLSKY_SFC_SW_DWN", {}).get("ANN")
                if ann and ann > 0:
                    # kWh/m²/day => approximate sunshine hours/year
                    return (lat, lon, round(ann * 365 * 0.75))
                return (lat, lon, None)
        except Exception:
            if attempt < 2:
                time.sleep(2 * (attempt + 1))
    return (lat, lon, None)


t0 = time.time()
results = []
done = 0

with ThreadPoolExecutor(max_workers=WORKERS) as pool:
    futures = {pool.submit(fetch_point, lat, lon): (lat, lon) for lat, lon in coords}

    for future in as_completed(futures):
        result = future.result()
        results.append(result)
        done += 1

        if done % 50 == 0 or done == total:
            pct = done / total * 100
            elapsed = time.time() - t0
            rate = done / elapsed if elapsed > 0 else 0
            eta = int((total - done) / rate) if rate > 0 else 0
            bar = "#" * int(pct / 100 * 40) + "." * (40 - int(pct / 100 * 40))
            sys.stdout.write(f"\r  {bar} {pct:.1f}% ({done}/{total}) ETA: {eta//60}m{eta%60}s   ")
            sys.stdout.flush()

print("\n")

# Sort by lat, lon for consistency
results.sort(key=lambda r: (r[0], r[1]))

valid = [r for r in results if r[2] is not None]
hours = [r[2] for r in valid]
mn, mx, avg = min(hours), max(hours), round(sum(hours) / len(hours))

output = {
    "step": STEP,
    "year": "2001-2020 climatology",
    "source": "NASA POWER (CERES SYN1deg / MERRA-2)",
    "stats": {"min": mn, "max": mx, "avg": avg, "count": len(valid), "total": len(results)},
    "grid": [[r[0], r[1], r[2]] for r in results],
}

OUT.parent.mkdir(parents=True, exist_ok=True)
OUT.write_text(json.dumps(output))

elapsed = (time.time() - t0) / 60
print(f"  OK Done! {len(valid)}/{len(results)} valid points")
print(f"  OK Range: {mn}-{mx} hours/year (avg: {avg})")
print(f"  OK Saved to {OUT}")
print(f"  OK Elapsed: {elapsed:.1f} minutes\n")
